Very long build configuration times when building BWC distributions

I noticed something very interesting today, it seems building BWC distributions can be _very_ slow when running something like `bwcTestSnapshots`. For example: https://gradle-enterprise.elastic.co/s/jgqlmj57k7s3e/timeline?task-path=buildBwcLinuxTar

8 minutes to build a distribution? That's odd, what's taking so long? Turns out, it's all [configuration time](https://gradle-enterprise.elastic.co/s/fhpzfrljafx7u/timeline). I've observed this locally as well, and in other BWC scenarios such as when doing `resolveAllDep`. My suspicion is the root cause here is some kind of lock contention within Gradle. Whether that's dependency or build script caches is unclear w/o doing thread dumps but I've seen issues where our `packer_cache.sh` script fails as well with errors waiting on locks.

This could be an issue in Gradle's optimistic locking. Daemons should be able to "request" these locks from eachother but it seems perhaps that's not working properly. Or it's locking at too granular a level and we're running into deadlocks. One workaround on our end would be to serialize these tasks w/ ordering rules. It's not ideal but it's better than having workers sitting around for 5 minutes or more doing nothing.

Here's an example of where I believe this lock contention is causing packer cache failures.

```
> Task :distribution:bwc:staged:resolveAllBwcDependencies
 [7.12.0] 
 [7.12.0] FAILURE: Build failed with an exception.
 [7.12.0] 
 [7.12.0] * What went wrong:
 [7.12.0] A problem occurred configuring project ':plugins:discovery-ec2'.
 [7.12.0] > Timeout waiting to lock jars (/var/lib/jenkins/.gradle/caches/jars-8). It is currently in use by another Gradle instance.
 [7.12.0]   Owner PID: unknown
 [7.12.0]   Our PID: 87508
 [7.12.0]   Owner Operation: unknown
 [7.12.0]   Our operation: 
 [7.12.0]   Lock file: /var/lib/jenkins/.gradle/caches/jars-8/jars-8.lock
 [7.12.0] > description is a required setting for esplugin
 [7.12.0] 
 [7.12.0] * Try:
 [7.12.0] Run with --info or --debug option to get more log output. Run with --scan to get full insights.
 [7.12.0] 
 [7.12.0] * Exception is:
 [7.12.0] org.gradle.api.ProjectConfigurationException: A problem occurred configuring project ':plugins:discovery-ec2'.
 [7.12.0] 	at org.gradle.configuration.project.LifecycleProjectEvaluator.wrapException(LifecycleProjectEvaluator.java:75)
 [7.12.0] 	at org.gradle.configuration.project.LifecycleProjectEvaluator.addConfigurationFailure(LifecycleProjectEvaluator.java:68)
 [7.12.0] 	at org.gradle.configuration.project.LifecycleProjectEvaluator.access$400(LifecycleProjectEvaluator.java:51)
 [7.12.0] 	at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.lambda$run$0(LifecycleProjectEvaluator.java:102)
 [7.12.0] 	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$0(DefaultProjectStateRegistry.java:250)
 [7.12.0] 	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$withProjectLock$3(DefaultProjectStateRegistry.java:310)
 [7.12.0] 	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:213)
 [7.12.0] 	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.withProjectLock(DefaultProjectStateRegistry.java:310)
 [7.12.0] 	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:291)
 [7.12.0] 	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:249)
 [7.12.0] 	at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:91)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
 [7.12.0] 	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
```

I'm wondering why the owner pid is showing "unknown" here. This implies that the inter-process communication that _should_ be used to ask other daemons to release this locks is not working.

--- a/buildSrc/src/main/java/org/elasticsearch/gradle/internal/BwcSetupExtension.java
+++ b/buildSrc/src/main/java/org/elasticsearch/gradle/internal/BwcSetupExtension.java
@@ -37,17 +37,20 @@ public class BwcSetupExtension {
 
     private static final String MINIMUM_COMPILER_VERSION_PATH = "buildSrc/src/main/resources/minimumCompilerVersion";
     private final Project project;
-
     private final Provider<BwcVersions.UnreleasedVersionInfo> unreleasedVersionInfo;
+    private final Provider<InternalDistributionBwcSetupPlugin.BwcTaskThrottle> bwcTaskThrottleProvider;
+
     private Provider<File> checkoutDir;
 
     public BwcSetupExtension(
         Project project,
         Provider<BwcVersions.UnreleasedVersionInfo> unreleasedVersionInfo,
+        Provider<InternalDistributionBwcSetupPlugin.BwcTaskThrottle> bwcTaskThrottleProvider,
         Provider<File> checkoutDir
     ) {
         this.project = project;
         this.unreleasedVersionInfo = unreleasedVersionInfo;
+        this.bwcTaskThrottleProvider = bwcTaskThrottleProvider;
         this.checkoutDir = checkoutDir;
     }
 
@@ -58,6 +61,7 @@ public class BwcSetupExtension {
     private TaskProvider<LoggedExec> createRunBwcGradleTask(Project project, String name, Action<LoggedExec> configAction) {
         return project.getTasks().register(name, LoggedExec.class, loggedExec -> {
             loggedExec.dependsOn("checkoutBwcBranch");
+            loggedExec.usesService(bwcTaskThrottleProvider);
             loggedExec.setSpoolOutput(true);
             loggedExec.setWorkingDir(checkoutDir.get());
             loggedExec.doFirst(t -> {

--- a/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
+++ b/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
@@ -205,7 +205,7 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
     private static String buildBwcTaskName(String projectName) {
         return "buildBwc"
             + stream(projectName.split("-")).map(i -> i.substring(0, 1).toUpperCase(Locale.ROOT) + i.substring(1))
-            .collect(Collectors.joining());
+                .collect(Collectors.joining());
     }
 
     static void createBuildBwcTask(
@@ -312,6 +312,5 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
         }
     }
 
-    public abstract class BwcTaskThrottle implements BuildService<BuildServiceParameters.None> {
-    }
+    public abstract class BwcTaskThrottle implements BuildService<BuildServiceParameters.None> {}
 }

--- a/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
+++ b/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
@@ -12,11 +12,14 @@ import org.elasticsearch.gradle.BwcVersions;
 import org.elasticsearch.gradle.Version;
 import org.elasticsearch.gradle.info.BuildParams;
 import org.elasticsearch.gradle.info.GlobalBuildInfoPlugin;
+import org.elasticsearch.gradle.util.GradleUtils;
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
 import org.gradle.api.tasks.TaskProvider;
 import org.gradle.language.base.plugins.LifecycleBasePlugin;
 
@@ -29,6 +32,7 @@ import java.util.stream.Collectors;
 
 import static java.util.Arrays.asList;
 import static java.util.Arrays.stream;
+import static org.elasticsearch.gradle.util.GradleUtils.noop;
 
 /**
  * We want to be able to do BWC tests for unreleased versions without relying on and waiting for snapshots.
@@ -39,6 +43,7 @@ import static java.util.Arrays.stream;
  */
 public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
 
+    private static final String BWC_TASK_THROTTLE_SERVICE = "bwcTaskThrottle";
     private ProviderFactory providerFactory;
 
     @Inject
@@ -49,19 +54,26 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
     @Override
     public void apply(Project project) {
         project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        Provider<BwcTaskThrottle> bwcTaskThrottleProvider = project.getGradle()
+            .getSharedServices()
+            .registerIfAbsent(BWC_TASK_THROTTLE_SERVICE, BwcTaskThrottle.class, spec -> spec.getMaxParallelUsages().set(1));
         BuildParams.getBwcVersions()
             .forPreviousUnreleased(
                 (BwcVersions.UnreleasedVersionInfo unreleasedVersion) -> {
-                    configureBwcProject(project.project(unreleasedVersion.gradleProjectPath), unreleasedVersion);
+                    configureBwcProject(project.project(unreleasedVersion.gradleProjectPath), unreleasedVersion, bwcTaskThrottleProvider);
                 }
             );
     }
 
-    private void configureBwcProject(Project project, BwcVersions.UnreleasedVersionInfo versionInfo) {
+    private void configureBwcProject(
+        Project project,
+        BwcVersions.UnreleasedVersionInfo versionInfo,
+        Provider<BwcTaskThrottle> bwcTaskThrottleProvider
+    ) {
         Provider<BwcVersions.UnreleasedVersionInfo> versionInfoProvider = providerFactory.provider(() -> versionInfo);
         Provider<File> checkoutDir = versionInfoProvider.map(info -> new File(project.getBuildDir(), "bwc/checkout-" + info.branch));
         BwcSetupExtension bwcSetupExtension = project.getExtensions()
-            .create("bwcSetup", BwcSetupExtension.class, project, versionInfoProvider, checkoutDir);
+            .create("bwcSetup", BwcSetupExtension.class, project, versionInfoProvider, bwcTaskThrottleProvider, checkoutDir);
         BwcGitExtension gitExtension = project.getPlugins().apply(InternalBwcGitPlugin.class).getGitExtension();
         Provider<Version> bwcVersion = versionInfoProvider.map(info -> info.version);
         gitExtension.setBwcVersion(versionInfoProvider.map(info -> info.version));
@@ -195,7 +207,7 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
     private static String buildBwcTaskName(String projectName) {
         return "buildBwc"
             + stream(projectName.split("-")).map(i -> i.substring(0, 1).toUpperCase(Locale.ROOT) + i.substring(1))
-                .collect(Collectors.joining());
+            .collect(Collectors.joining());
     }
 
     static void createBuildBwcTask(
@@ -252,7 +264,7 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
         /**
          * can be removed once we don't build 7.10 anymore
          * from source for bwc tests.
-         * */
+         */
         @Deprecated
         final boolean expandedDistDirSupport;
         final DistributionProjectArtifact expectedBuildArtifact;
@@ -286,7 +298,7 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
         /**
          * Newer elasticsearch branches allow building extracted bwc elasticsearch versions
          * from source without the overhead of creating an archive by using assembleExtracted instead of assemble.
-         * */
+         */
         public String getAssembleTaskName() {
             return extractedAssembleSupported ? "extractedAssemble" : "assemble";
         }
@@ -300,5 +312,8 @@ public class InternalDistributionBwcSetupPlugin implements InternalPlugin {
             this.distFile = distFile;
             this.expandedDistDir = expandedDistDir;
         }
+    }
+
+    public abstract class BwcTaskThrottle implements BuildService<BuildServiceParameters.None> {
     }
 }

--- a/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
+++ b/buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java
@@ -12,7 +12,6 @@ import org.elasticsearch.gradle.BwcVersions;
 import org.elasticsearch.gradle.Version;
 import org.elasticsearch.gradle.info.BuildParams;
 import org.elasticsearch.gradle.info.GlobalBuildInfoPlugin;
-import org.elasticsearch.gradle.util.GradleUtils;
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
@@ -32,7 +31,6 @@ import java.util.stream.Collectors;
 
 import static java.util.Arrays.asList;
 import static java.util.Arrays.stream;
-import static org.elasticsearch.gradle.util.GradleUtils.noop;
 
 /**
  * We want to be able to do BWC tests for unreleased versions without relying on and waiting for snapshots.

